src/sgame/sg_votes.cpp

```diff
@@ -0,0 +1,1340 @@
+                                         std::string& name, int clientNum, int id, bool escape )
+{
+	std::unordered_map<std::string, std::string> params = {
+		{ "team", BG_TeamNamePlural( team ) },
+		{ "arg", VoteEscape( arg, escape ) },
+		{ "reason", VoteEscape( reason, escape ) },
+		{ "name", VoteEscape( name, escape ) },
+		{ "slot", std::to_string( clientNum ) },
+		{ "namelogId", std::to_string( id ) },
+		{ "caller", VoteEscape( ent->client->pers.netname, escape ) },
+	};
+	std::string out;
+	out.reserve( str.size() + arg.size() + reason.size() );
+	size_t c = 0;
+	size_t s = 0;
+	size_t e = 0;
+	while ( c < str.size() )
+	{
+		s = str.find( MARKER, c );
+		if ( s != std::string::npos && !IsEscapedMarker( str, s ) )
+		{
+			e = str.find( MARKER, s + 1 );
+			if ( e != std::string::npos )
+			{
+				out += str.substr( c, s - c );
+				out += params[ str.substr( s + 1, e - s - 1 ).c_str() ];
+				c = e + 1;
+			}
+			else
+			{
+				Log::Warn( "Unterminated %c in str: %s", MARKER, str );
+				return "";
+			}
+		}
+		else
+		{
+			out += str.substr( c );
+			c = str.size();
+		}
+	}
+	Log::Notice( " -- %s", out.c_str() );
+	return out;
+}
+
+bool G_AddCustomVote( std::string vote, VoteDefinition def, std::string voteTemplate,
+                      std::string displayTemplate )
+{
+	if ( vote.empty() )
+	{
+		return false;
+	}
+	auto it = voteInfo.find( vote );
+	if ( it != voteInfo.end() )
+	{
+		return false;
+	}
+	it = voteInfo.emplace( std::move( vote ), std::move( def ) ).first;
+	auto handler = [ vt = move( voteTemplate ), dt = move( displayTemplate ) ](
+					   gentity_t* ent, team_t team, std::string& cmd, std::string& arg,
+					   std::string& reason, std::string& name, int clientNum, int id )
+	{
+		std::string voteCmd =
+			G_HandleVoteTemplate( vt, ent, team, cmd, arg, reason, name, clientNum, id, true );
+		Q_strncpyz( level.team[ team ].voteString, voteCmd.c_str(),
+		            sizeof( level.team[ team ].voteString ) );
+		std::string display =
+			G_HandleVoteTemplate( dt, ent, team, cmd, arg, reason, name, clientNum, id, false );
+		Q_strncpyz( level.team[ team ].voteDisplayString, display.c_str(),
+		            sizeof( level.team[ team ].voteDisplayString ) );
+		return true;
+	};
+	it->second.handler = std::move( handler );
+	return true;
+}
+
+bool ParseVoteType( Str::StringRef s, VoteType* type )
+{
+	static const std::unordered_map<std::string, VoteType> map = {
+		{ "V_ANY", V_ANY },
+		{ "V_TEAM", V_TEAM },
+		{ "V_PUBLIC", V_PUBLIC },
+	};
+	const auto& it = map.find( s );
+	if ( it == map.end() )
+	{
+		return false;
+	}
+	*type = it->second;
+	return true;
+}
+
+Str::StringRef VoteTypeString( VoteType type )
+{
+	switch ( type )
+	{
+		case V_TEAM:
+			return "V_TEAM";
+		case V_PUBLIC:
+			return "V_PUBLIC";
+		case V_ANY:
+		default:
+			return "V_ANY";
+	}
+}
+
+bool ParseVoteTarget( Str::StringRef s, VoteTarget* type )
+{
+	static const std::unordered_map<std::string, VoteTarget> map = {
```

Comment:
We usually use case-insensitive parsing. `std::unordered_map<std::string, VoteTarget, Str::IHash, Str::IEqual>`

Reply:
Done